[SPIRV] Add support for lowering specified in IR

harsh-nod opened the pull request:
This patch adds support to have config coming from higher level IR (to enable search).

It reuses the functionality from the LLVMGPU pipeline.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/Transforms.h:
```diff
@@ -19,5 +20,9 @@ void populateTileAndDistributeToWorkgroupsPatterns(
 /// hal.interface.binding.subspan.
 void populateReshapeToInterfaceTensorPatterns(RewritePatternSet &patterns);
 
+/// Sets compilation configuration annotated in the incoming IR.
```
Could we move this to `Utils`?

harsh-nod replied:
sure

harsh-nod replied:
Regarding moving this to Utils, this function requires IREE::Codegen::CompilationInfoAttr which is part of the IREECodegenDialect but that library requires the Utils lib so we end up with a circular dependency. So I don't think that will work. Should I keep it in Common or do you have any suggestions on how to get around the circular dependency?

MaheshRavishankar reviewed (changes requested):
Do you mind using this here as well https://github.com/iree-org/iree/blob/bc38c26e9c8da33f80c3cac970909c342b527de3/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1520 (I think its almost the exact same code)

harsh-nod commented:
> Do you mind using this here as well
> 
> https://github.com/iree-org/iree/blob/bc38c26e9c8da33f80c3cac970909c342b527de3/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1520
> 
> (I think its almost the exact same code)

yeah looks almost identical. sure will change.

harsh-nod commented:
@MaheshRavishankar @antiagainst @ThomasRaoux - I have updated the patch to pass all the checks. What I noticed for the e2e matmul tests for vulkan (for NVIDIA and Android) was that the tile sizes supplied by the test were not being used (this was causing the test to fail for my patch). I fixed this by specifying the actual tile sizes being used (I found the actual tile sizes by looking at the IR after the tile and distribute pass). Please take a look and let me know what you think. Thanks!

antiagainst commented on tests/e2e/matmul/generate_e2e_matmul_tests.py:
```diff
@@ -155,7 +157,7 @@ def get_dynamicities(shapes_id: ShapesId):
 @dataclasses.dataclass
 class TileWorkgroupSizePair:
-  tile_size: typing.List[int]
+  tile_size: typing.Union[typing.List[int], typing.List[typing.List[int]]]
```
Instead of doing this union and the if-else cases in the following, let's change the definition of `TileWorkgroupSizePair` to use `typing.List[typing.List[int]]` for `tile_size` and update other cases. This would avoid quite some complexity.

harsh-nod replied:
Sure sounds good. I have updated the python file.

antiagainst commented on tests/e2e/matmul/generate_e2e_matmul_tests.py:
```diff
@@ -176,6 +177,26 @@ def get_test_compilation_infos(
     ]
+  elif compilation_info_id == CompilationInfoId.SPIRVVectorizeNVIDIA:
+    tile_workgroup_size_pairs = [
+        TileWorkgroupSizePair([[32, 128], [4, 4], [0, 0, 32]], [32, 8, 1]),
```
Hmm, these numbers are inconsistent. Say if we have the following

```
[[wg_tile_m, wg_tile_n], [t_tile_m, t_tile_n], [0, 0, tile_k]], [x, y, z]
```
then `wg_tile_m = y * t_thie_m` and `wg_tile_n = x * t_tile_n`. So we need to fix the first tile size list in the tuple.

antiagainst replied:
I think another way to move SPIR-V side to only have one tile size like what we do for CUDA side. But that can happen separately as another patch.

harsh-nod replied:
Thanks. I was not aware of that constraint, but I have added it to the test so that all future sizes will obey that constraint. Yes agreed on the single tile size for a later patch.

antiagainst commented:
@harsh-nod thanks a lot for pushing on this! Yeah previously we don't plumb user specified lowering configurations down the pipeline so under the hood it's still the heuristics deciding tile sizes and workgroup sizes. Now you are connecting this so it broke. Thanks for fixing it!

harsh-nod commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1519,19 +1520,7 @@ static LogicalResult setTranslationInfoAndRootConfig(
-                         workgroupSize);
-      setLoweringConfig(computeOp, compilationInfo.getLoweringConfig());
-      eraseCompilationInfo(computeOp);
+      (void)setUserConfig(entryPointFn, computeOp, compilationInfo);
```
sounds good. I have updated the patch.

harsh-nod commented on compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp:
```diff
@@ -684,7 +685,15 @@ static LogicalResult setDefaultOpConfig(spirv::ResourceLimitsAttr limits,
                                       Operation *rootOp) {
+  if (IREE::Codegen::CompilationInfoAttr compilationInfo =
+          getCompilationInfo(rootOp)) {
+    // If the op already has a lowering config coming from the IR use this and
```
I have updated the comment.

harsh-nod commented on compiler/src/iree/compiler/Codegen/SPIRV/test/config_user.mlir:
```diff
@@ -0,0 +1,55 @@
+// RUN: iree-opt --split-input-file --pass-pipeline='hal.executable(hal.executable.variant(iree-spirv-lower-executable-target-pass{test-lowering-configuration=true}))' %s | FileCheck %s
+
+#compilation = #iree_codegen.compilation_info<
```
Makes sense. I wasn't sure what numbers to use but this helps. I have updated the test with these parameters.

harsh-nod commented on compiler/src/iree/compiler/Codegen/SPIRV/test/config_user.mlir:
```diff
@@ -0,0 +1,55 @@
+
+#compilation = #iree_codegen.compilation_info<
+    lowering_config  = <tile_sizes = [[8, 8], [1, 1]]>,
+    translation_info = <SPIRVVectorize workload_per_wg = [64, 8]>,
```
No it is not needed. I have removed it from the patch.

harsh-nod commented on tests/e2e/matmul/BUILD:
```diff
@@ -257,9 +257,9 @@ py_binary(
-    "valhall-unknown-android31",
-    "ampere-unknown-linux",
+) for vulkan_target_and_pipeline in [
+    ("valhall-unknown-android31", "SPIRVVectorizeAndroid"),
```
I have renamed it in the python and BUILD file.

harsh-nod commented:
> @harsh-nod thanks a lot for pushing on this! Yeah previously we don't plumb user specified lowering configurations down the pipeline so under the hood it's still the heuristics deciding tile sizes and workgroup sizes. Now you are connecting this so it broke. Thanks for fixing it!

Thank you for all your comments and feedback!

MaheshRavishankar reviewed (approved):
Thanks @harsh-nod

antiagainst commented:
Nice, thanks @harsh-nod!

harsh-nod commented:
Thanks @MaheshRavishankar @antiagainst for guiding me through this! :)
